### PR TITLE
Swift extensions to Vokoder

### PR DESCRIPTION
**Pull request by chillpop**
Added some swift-ier extensions to VOKCoreDataManager and the NSManagedObject category functions. The extensions return typed subclasses instead of `NSManagedObject` to allow for less casting in swift code.  Also, some nullable parameters now have default values for cleaner function calls.

All the swift code is in an optional subspec of the main pod so it is not included by default.

The swift example project has been updated to use the new extensions.

@vokal/ios-developers, @seanwolter, code review?

### CODE REVIEW

**Comment by chillpop** [comment]
The diff count looks bad, but there's not a lot of code to review.  It's a lot of a CTA data file (not shown) and a pod project file update for the swift example project.

**Comment by designatednerd** [comment]
oooooOOOOOOOooooo

(on it)

File: Pod/Classes/Vokoder.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+//  Copyright © 2016 Vokal.
+//
+
+#ifndef Vokoder_h
+#define Vokoder_h
+
+//Xcode assumes a header file named the same as the module exists and includes it
+//in the autogenerated swift interface to Obj-C classes.
+//To allow for mixing Obj-C and Swift in the Vokoder framework,
+//we create this empty file to satisfy the compiler.
```
Comment (designatednerd) [inline Pod/Classes/Vokoder.h]: Good explanation. Silly compiler.

File: SwiftSampleProject/SwiftyVokoder/Models/Station.swift [hunk 2]
```diff
@@ -70,3 +70,5 @@ extension Station: VOKMappableModel {
         }
     }
 }
+
+extension Station: VokoderTypedManagedObject { }
```
Comment (designatednerd) [inline SwiftSampleProject/SwiftyVokoder/Models/Station.swift]: a) Needs moar `//MARK`
b) Since this is a sample project, is it worth throwing in any kind of note about the fact that this (and the several below) are basically mixin declarations?

File: SwiftSampleProject/SwiftyVokoderTests/SwiftyVokoderTests.swift [hunk 3]
```diff
@@ -65,4 +94,18 @@ class SwiftyVokoderTests: XCTestCase {
...
+        
+        XCTAssertEqual(stops.count, 300)
+        guard let stop = stops.filter({ stop in
+            return stop.identifier == testIdentifier
+        }).first else {
+            XCTFail("Could not find stop \(testIdentifier)")
+            return
+        }
+        
+        self.verifyGrandMilwaukeeStop(stop)
```
Comment (designatednerd) [inline SwiftSampleProject/SwiftyVokoderTests/SwiftyVokoderTests.swift]: I assume the stop you're looking for is the test identifier stop? Might be helpful to change the name of `testIdentifier` to `grandMilwaukeeIdentifier` or the name of this method to `verifyStopWithTestIdentifier` just to make that clearer.

File: Pod/Classes/Swift/VOKCoreDataManager+Swift.swift [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+     - parameter context: The managed object context in which to create the objects or nil for the main context (defaults to nil)
+     - returns: A typed Array of created or updated objects
+     */
+    public func importArray<T: NSManagedObject>(inputArray: [[String : AnyObject]],
+        forClass objectClass: T.Type,
+        withContext context: NSManagedObjectContext? = nil) -> [T] {
+            guard let result = self.importArray(inputArray,
+                forClass: objectClass,
+                withContext: context) as? [T] else {
+                    fatalError("Could not cast array of NSManagedObjects into \(String(T))")
```
Comment (designatednerd) [inline Pod/Classes/Swift/VOKCoreDataManager+Swift.swift]: Is there any way to add tests around this to see what happens if the import fails? If it doesn't return an empty array, It might be useful to have this throw an error rather than just dying in that case.

File: Pod/Classes/Swift/VokoderTypedManagedObject.swift [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+     NOT threadsafe! Always use a temp context if you are NOT on the main queue.
+     
+     - parameter predicate: The predicate limit the fetch (defaults to nil)
+     - parameter sortedBy: The sort descriptors to sort the results (defaults to nil)
+     - parameter context: The managed object context in which to fetch objects or nil for the main context (defaults to nil)
+     - returns: A typed Array of managed object subclasses. Not threadsafe.
+     */
+    public static func vok_fetchAll(forPredicate predicate: NSPredicate? = nil,
+        sortedBy sortDescriptors: [NSSortDescriptor]? = nil,
+        forManagedObjectContext context: NSManagedObjectContext? = nil) -> [Self] {
```
Comment (designatednerd) [inline Pod/Classes/Swift/VokoderTypedManagedObject.swift]: Worth adding tests to ensure these things return the same thing just for future-proofing purposes? Basically, if someone else has the bright idea to alter this, it'd help to know immediately when it breaks.

File: Pod/Classes/Swift/VOKCoreDataManager+Swift.swift [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    public func arrayForClass<T: NSManagedObject>(objectClass: T.Type,
+        withPredicate predicate: NSPredicate? = nil,
+        sortedBy sortDescriptors: [NSSortDescriptor]? = nil,
+        forContext context: NSManagedObjectContext? = nil) -> [T] {
+            guard let result = self.arrayForClass(objectClass,
+                withPredicate: predicate,
+                sortedBy: sortDescriptors,
+                forContext: context) as? [T] else {
+                    fatalError("Could not cast array of NSManagedObjects into \(String(T))")
```
Comment (designatednerd) [inline Pod/Classes/Swift/VOKCoreDataManager+Swift.swift]: Does this return an empty array of `T`, or does it fatal error if there are no objects of the given class?

File: SwiftSampleProject/SwiftyVokoder/Models/Station.swift (hunk 2 above)
Comment (designatednerd) [inline SwiftSampleProject/SwiftyVokoder/Models/Station.swift]: (I see the note in the actual protocol, but it might not hurt to put something in with the code as well to be 100% clear)

File: Vokoder.podspec [hunk 7]
```diff
@@ -41,4 +43,11 @@ Pod::Spec.new do |s|
       sss.dependency 'Vokoder/DataSources/FetchedResults'
     end
   end
+
+  s.subspec 'Swift' do |sw|
```
Comment (designatednerd) [inline Vokoder.podspec]: :notes: Come on baby, let's do the Swift :notes:

**Comment by chillpop** [comment]
updates pushed

File: Pod/Classes/Swift/VOKCoreDataManager+Swift.swift (hunk 6 above)
Comment (chillpop) [inline Pod/Classes/Swift/VOKCoreDataManager+Swift.swift]: empty array, (per the added test)

File: Pod/Classes/Swift/VokoderTypedManagedObject.swift (hunk 5 above)
Comment (chillpop) [inline Pod/Classes/Swift/VokoderTypedManagedObject.swift]: test added

**Comment by designatednerd** [comment]
LGTM. Anyone else wanna take a crack at this?

**Comment by brockboland** [comment]
I'll take a gander. I wanted to this morning but got sidetracked.

**Comment by brockboland** [comment]
LGTM2. If anyone else wants to take a look, please speak up by EOD.
